### PR TITLE
Add Equal and Not Equal tests for Tuples with null fields.

### PR DESCRIPTION
**Pull request by suyashkumar**
This change adds equal and not equal tests for Tuples that have missing values, based on clarifications and discussions in the [Zulip Chat](https://chat.fhir.org/#narrow/stream/179220-cql/topic/NotEquivalent.20ELM.20representation/near/425401720).

We may wish to make a change to add cases like this for Equivalent and Not Equivalent too, but I limited this PR to the discussion at hand.

__Questions__:
* We don't yet have any CI for these tests yet, so what testing or validation do we wish to do with these before submitting? I have manually run these in the Java engine successfully (except for ones where Bryn said the Java behavior is incorrect). 
* I wasn't sure about the test naming convention, so took a random stab at it.

### CODE REVIEW

File: tests/cql/CqlComparisonOperatorsTest.xml
Comment (brynrhodes) [inline tests/cql/CqlComparisonOperatorsTest.xml:98]: This should be `null`, tuples have no implied ordering, so this is a bug in the Java engine...

File: tests/cql/CqlComparisonOperatorsTest.xml
Comment (brynrhodes) [inline tests/cql/CqlComparisonOperatorsTest.xml:780]: Same here, this should be `null`

**Review by brynrhodes — CHANGES_REQUESTED** [review]

File: tests/cql/CqlComparisonOperatorsTest.xml
Comment (suyashkumar) [reply tests/cql/CqlComparisonOperatorsTest.xml:98]: Ah interesting! I was thinking that null is used to indicate "we don't know if these are equal or not" but in this case, since two of the set fields (Id) are unequal I expected that we could confidently return false, regardless of the status of the other fields. Is this not the case?

Comment (brynrhodes) [reply tests/cql/CqlComparisonOperatorsTest.xml:98]: There's room for interpretation here, but the spec says "and the values for all elements that have values, by name, are equal.", to me, that means the element in question has a value in at least one of the inputs. But I can see it being interpreted as "elements that have values for both inputs", and would allow us to return false here. What I'd be concerned with is making sure that didn't introduce an ordering dependency in tuple comparison, so if we did say it was false, then this should return false too:
```cql
{ Id: null, Name: 'John' } = { Id: null, Name: 'Joe' }
```
And indeed, that is how the Java engine is behaving...

File: tests/cql/CqlComparisonOperatorsTest.xml
Comment (brynrhodes) [reply tests/cql/CqlComparisonOperatorsTest.xml:780]: And as long as we always ensured that the `!=` was equivalent to a `!` of `=`, then that should be okay.

File: tests/cql/CqlComparisonOperatorsTest.xml
Comment (brynrhodes) [reply tests/cql/CqlComparisonOperatorsTest.xml:98]: And even if we were to long-hand it as a conjunction of ands... that would still return false by 3VL semantics for and. So yes, I think you're right, and the tests as written here are correct. (And the Java engine is behaving correctly). Can we add this test though to make sure we're not introducing an ordering dependency?

Comment (suyashkumar) [reply tests/cql/CqlComparisonOperatorsTest.xml:98]: +1 agreed my 2 cents are this behavior would be nice, as it returns false when two tuples are knowably false!

Makes sense on the ordering dependency. For the Java engine, testing the following actually _does_ return null (with one of the Ids being set and the other unset), so perhaps there is an ordering dep in the Java engine. 

Differing names with one null ID:
```
{ Id: null, Name: 'John' } = { Id: 1, Name: 'Joe' } = null // Java
```

Different IDs with one null name:
```
{ Id: 2, Name: 'John' } = { Id: 1, Name: null } = false // Java
```

So it seems in the current Java implementation there may indeed be a an ordering dependency. 

Current Java engine behavior aside, should we still go ahead with the convention that if any of the fields which are set on both operands are not equal, we should return false? Can add suggested tests cases if we're cool with that! This will mean that engines may need to compare more tuple fields in some cases to make a determination instead of returning null in a greedy fashion. (But false can still be returned in a greedy manner)!

Comment (brynrhodes) [reply tests/cql/CqlComparisonOperatorsTest.xml:98]: Yeah, good catch, there is some ordering dependency there, but yes, I agree we should define it as conjunction of equality of all elements, and an implementation can then be greedy about a false result.

Comment (suyashkumar) [reply tests/cql/CqlComparisonOperatorsTest.xml:98]: Fantastic, sounds good! Added some tests for the order swapped fields.

File: tests/cql/CqlComparisonOperatorsTest.xml
Comment (suyashkumar) [reply tests/cql/CqlComparisonOperatorsTest.xml:788]: Resolving as per the thread above, let me know if there was anything else needed here.

**Review by suyashkumar — COMMENTED** [review]
Thanks!

**Review by brynrhodes — APPROVED** [review]

**Comment by suyashkumar** [comment]
@JPercival @EvanMachusakNCQA if y'all have a moment to give a second review in addition to Bryn's, that'd be great--thanks!

**Review by EvanMachusakNCQA — APPROVED** [review]

**Pull request closed without merging**
